add doc around dashboard telemetry (#2846)
dotnet/docs-aspire: docs/fundamentals/dashboard/usage-telemetry.md

---
title: .NET Aspire dashboard telemetry
description: Learn about what telemetry the .NET Aspire dashboard sends and how to opt out.
ms.date: 03/05/2025
---

> # .NET Aspire dashboard telemetry

Reviewer: What do you think about this for the title instead?

```suggestion
# Microsoft collected dashboard telemetry
```

".NET Aspire telemetry" is now officially an oxymoron; in that it has two contradictory meanings. Before 9.2, the dashboard was all about display resource telemetry, obviously we all know that, but now, we're saying ".NET Aspire telemetry" in this article is about the data Microsoft collects to improve the dashboard experience.

Or perhaps something like ".NET Aspire dashboard usage telemetry"?

Author: Microsoft collected dashboard telemetry works for me. This wording also makes it easier if in the future additional telemetry is added to Aspire more broadly beyond the dashboard

> The Aspire dashboard includes a telemetry feature that collects usage data when the dashboard is launched through Visual Studio or Visual Studio Code as part of a running Aspire application. This information is sent to Microsoft to help the Aspire team understand how the dashboard is used and help improve the product. Exception information is also sent when unhandled exceptions occur in the dashboard.

Reviewer: In docs, it's always ".NET Aspire" per marketing - just an FYI. Also, I'm suggesting some changes to the text here as to avoid conflating .NET Aspire telemetry and .NET Aspire telemetry 🙃. As you know, the dashboard is all about telemetry, we need to be very specific when we describe it to avoid confusion.

```suggestion
The .NET Aspire dashboard collects diagnostic data to help developers monitor and analyze their applications. Separately, when the dashboard is launched through Visual Studio or Visual Studio Code as part of a running .NET Aspire application, Microsoft collects usage data about the dashboard itself. This data helps the .NET Aspire team improve the product. Additionally, unhandled exception details from the dashboard are sent to Microsoft to assist in identifying and resolving issues.
```

## Scope

> .NET Aspire dashboard usage telemetry is supported when using Visual Studio or Visual Studio Code to run an Aspire application containing a dashboard resource.
> 
> Telemetry is collected only when the Aspire dashboard is open in the browser and the instance of Visual Studio or Visual Studio Code has not opted out of telemetry collection.

Reviewer: ```suggestion
.NET Aspire dashboard usage telemetry is collected only when running the dashboard from Visual Studio or Visual Studio Code. Telemetry is gathered while the dashboard is open in the browser, unless you have [opted out](#how-to-opt-out) of telemetry collection.
```

## How to opt out

> .NET Aspire dashboard telemetry is enabled by default for `Aspire >= 9.2` when launched through Visual Studio `>= 17.14` or C# Dev Kit `>= [VSC RELEASE VERSION]`.

Reviewer: ```suggestion
Starting with .NET Aspire 9.2, dashboard usage telemetry is enabled by default. Dashboard usage telemetry is not collected in early version. This feature aligns with the following IDE versions:

- Visual Studio: 17.14 or later.
- C# Dev Kit: [VSC RELEASE VERSION] or later.
```

To opt out, you may either:

> - Set the `DOTNET_DASHBOARD_ENABLE_TELEMETRY` environment variable to `false`. This will apply to all users accessing the Aspire dashboard.

Reviewer: Could a user alternatively set an `AppContext` switch?

```cs
AppContext.SetSwitch("DOTNET_DASHBOARD_ENABLE_TELEMETRY", true);
```

Author: Not currently. That pattern is not currently supported for any dashboard configuration option, so it needs further discussion. @davidfowl how do you feel about supporting switches?

> - Disable telemetry collection in the host IDE.

Reviewer: How does one disable telemetry collection in the host IDE?

## Disclosure

> If dashboard telemetry is enabled, there will be a disclosure statement at the bottom of the settings panel informing that telemetry is enabled. This statement will appear any time telemetry collection is enabled.

Reviewer: You're using "any time" here to mean whenever, which is adverbial — so "anytime" is the correct form. Also, updated this to be active voice to align with docs style/tone.

```suggestion
When dashboard usage telemetry is enabled, there's a disclosure statement at the bottom of the settings panel, informing you that telemetry is enabled. This disclosure statement appears anytime usage telemetry is enabled.
```

:::image type="content" source="media/explore/dashboard-settings-drawer.png" lightbox="media/explore/dashboard-settings-drawer.png" alt-text="A screenshot of the .NET Aspire dashboard settings drawer.":::

## Data points

> Aspire dashboard telemetry does not collect personal data like IP addresses or use browser fingerprinting. It does not scan your code and does not extract source code, authorship, or deployment configuration. The data is sent securely to Microsoft using [https://azure.microsoft.com/services/monitor/](Azure Monitor) through existing telemetry APIs in Visual Studio and Visual Studio Code.

Reviewer: ```suggestion
.NET Aspire dashboard usage telemetry doesn't collect personal data, such as, IP addresses or use browser fingerprinting. It doesn't scan your code and doesn't extract source code, authorship, or deployment configuration. The data is sent securely to Microsoft using [https://azure.microsoft.com/services/monitor/](Azure Monitor) through existing telemetry APIs in Visual Studio and Visual Studio Code.
```

> Protecting your privacy is important to us. If you suspect that telemetry is collecting sensitive data or the data is being insecurely or inappropriately handled, file an issue in the [dotnet/aspire](https://github.com/dotnet/aspire) repository for investigation.

Reviewer: ```suggestion
Protecting your privacy is important to Microsoft. If you suspect that telemetry is collecting sensitive data or the data is being insecurely or inappropriately handled, file an issue in the [:::image type="icon" source="../../media/github-mark.svg border="false"::: GitHub dotnet/aspire](https://github.com/dotnet/aspire) repository for investigation.
```

> The Aspire dashboard does not collect telemetry on Visual Studio versions `< 17.14` or C# Dev Kit versions `< [VSC RELEASE VERSION`. It collects the following data:

Reviewer: ```suggestion
The .NET Aspire dashboard doesn't collect telemetry on Visual Studio versions `< 17.14` or C# Dev Kit versions `< [VSC RELEASE VERSION]`. It collects the following data:
```

> | Aspire dashboard versions | Data | Notes |
> |--------------|------|------|
> | 9.2          | Page navigation history. | Includes page settings. |
> | 9.3          | Resource types being used. | |
> | 9.3          | Request user agent. | |
> | 9.3          | Invoked dashboard commands. | Command name is securely hashed. |
> | 9.3          | Request language and set dashboard language. | |
> | 9.3          | Resource restart times. | |
> | 9.3          | OTel data processing times. | |
> | 9.3          | Dashboard-related unhandled exceptions. | |

Reviewer: ```suggestion
| .NET Aspire dashboard versions | Data | Notes |
|--|--|--|
| 9.2 | Page navigation history. | Includes page settings. |
| 9.3 | Resource types being used. |  |
| 9.3 | Request user agent. |  |
| 9.3 | Invoked dashboard commands. | Command name is securely hashed. |
| 9.3 | Request language and set dashboard language. |  |
| 9.3 | Resource restart times. |  |
| 9.3 | OTel data processing times. |  |
| 9.3 | Dashboard-related unhandled exceptions. |  |

## See also

- [.NET Aspire dashboard](overview.md)
- [.NET Aspire dashboard configuration](configuration.md)

```
